### ISSUE TITLE
Port twisted.web.client.downloadPage to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6197]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6197|
|Type|enhancement|
|Created|2012-11-29 19:41:23Z|
|Branch|https://github.com/twisted/twisted/tree/downloadpage-py3-6197-2|

`Port twisted.web.client.downloadPage` and related code should be ported to Python 3 (unless we decide to deprecate it before the port happens).

Attachments:

* [downloadPage.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6e4/6e45c36b27587802389b2721a4b445181e51ec81/84ac9229dbb89f14113e6dd79067463966409edb.patch) (1259 bytes) - added by vmarkovtsev on 2014-05-26 07:16:17Z - 
* [downloadPage_getPage.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6e4/6e45c36b27587802389b2721a4b445181e51ec81/5840b51ed6e9d9604856152e712ae766dcf9a8b9.patch) (1679 bytes) - added by vmarkovtsev on 2014-05-27 13:46:42Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6197 6197
type__enhancement enhancement
reporter__itamar itamar
priority__lowest lowest
milestone__Python_3_x Python-3.x
branch__branches_downloadpage_py3_6197_2 branches/downloadpage-py3-6197-2
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1354218083000000 1354218083000000
changetime__1445305364925960 1445305364925960
version__None None
owner__hawkowl hawkowl
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|vmarkovtsev commented|
|-|-|

My patch does it. It depends on porting web.client.Agent (ticket [#7407](https://github.com/twisted/twisted/issues/7407)).

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|vmarkovtsev commented|
|-|-|

web/test/test_webclient.py is passed 100% without any modifications

**Comment by twisted-trac** [comment]
|<img alt="vmarkovtsev's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|vmarkovtsev commented|
|-|-|

The second patch ports getPage() together with downloadPage().

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [45480]) Branching to downloadpage-py3-6197.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Tests are green. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Sorry for the delay.

I got some feedback from the ML and since Glyph is also in favor of bytes we no longer have blocking tie :)

Many thanks for the patience and for working on this.

---------

In HTTPDownloader is it ok to change the default value for arguments?
I know that 0/1 were supposed to be used as bool, but maybe some people were already using them as int and then the code will break.

For `isinstance(fileOrName, (str, unicode))` maybe we can implement in t.python.compat something like
isText(fileOrName) so that it make more sense and hint about the reason why we check the type.

In twisted/web/error.py why do we need to convert the code to bytes when it is documented as requiring bytes?

In twisted/web/test/test_webclient.py do we still need to encode('utf-8') the test.__file__ ?

-------

Please check my comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl removed owner|
|-|-|

Hi Adi, thanks for the review.

1. I think it's fine. It was quite clearly True/False before, so it's better to have it explicitly so now.
2. I think that such a general check isn't that good -- this is just replacing what ```types.StringTypes``` was, and it shouldn't be replicated in other errors.
3. This is because some things gave it as an int, but that was only tests. Fixed.
4. No, we don't. Fixed.

Builders spun. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Hi,

Thanks for the update!

There is no mention in this ticket or in the code/news fragment why int is now an accepted type for the argument.

In a previous comments you mentioned that it should only accept bytes. Why does it still need to accept int?

Please check my comments and resubmit.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|posita commented|
|-|-|

Two notes:

1. This might be the same question as @adiroiban's most recent, but if I'm reading your prior response correctly, namely ...

    >This is because some things gave it as an int, but that was only tests. Fixed.

  ... then the `if isinstance(code, int)` block should no longer be necessary in `t.w.e.Error.__init__`. Does it need to remain for backward compatibility with third party libraries?

2. This is a _super_ minor nitpick (feel free to ignore).

  In the docstrings for some places (see t.w.e.Error as an example), `C{str}` is changed to `C{bytes}`, but in others, `C{str}` (or `C{bytes}`, etc.) is changed to `L{bytes}`. Does `L{...}` make sense with builtin types?

  _Some_ of these look like they're adjusted to `C{bytes}` in [45485] (see #7404), but not all.

  This is probably inconsequential. It appears `L{<builtin-type>}` renders the same as `C{<builtin-type}`, at least in all the places I can see.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

(In [45786]) Branching to downloadpage-py3-6197-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_12"></a>|@hawkowl removed owner|
|-|-|

Okay, I've added a note about backwards compatibility.

RE: posita -- ```L{```} does make sense for built in types -- ```L{```} should link to the Python API docs, now or in the future. If it can't resolve it, it works the same as ```C{```}.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Since the `code` is now documented as both bytes and int I think that the conversion code is not there for backward compatibility but to comply with the docstring.

If we want to keep backward compatiblity, maybe we can silently do the conversion, without also advertising the support for `int` in the docstring.

Also, from I understand, the problem is with the Agent code which passes 'int' instead of 'str' (as it was documented).
Since downloadPage passed str and not bytes, it should be fine and it does not require any special backward compatibility.

Please update the docstring or the comment... I am still confused :(

Thanks for the changes!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Replying to [adiroiban](#note_14):
> Since the `code` is now documented as both bytes and int I think that the conversion code is not there for backward compatibility but to comply with the docstring.

The comment doesn't actually _say_ "for backwards compatibility", it just gives some historical context.

> If we want to keep backward compatiblity, maybe we can silently do the conversion, without also advertising the support for `int` in the docstring.

The problem with not advertising `int` in the docstring is that all of our public constants in `twisted.web.http` are exposed as `int`.  So that would be a glaring incompatibility in the API.  But, HTTP responses may received malformed codes, so sometimes it makes sense to get the bytes instead, since it may not fit into an `int`.  Although IANA has never assigned something which cannot be represented as an int - http://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml - microsoft has a couple of weird extensions - (see 401.1, etc, here: https://support.microsoft.com/en-us/kb/943891 )  I do not think it is very clear-cut what to do in this case.

> Also, from I understand, the problem is with the Agent code which passes 'int' instead of 'str' (as it was documented).
> Since downloadPage passed str and not bytes, it should be fine and it does not require any special backward compatibility.

`str` and `int` are still two quite different types; I'm not sure what you're getting at here.

> Please update the docstring or the comment... I am still confused :(

I would replace the `/` with the word `or`, but other than that, I actually think both docstring and comment are fine.  If we want a holistic redesign of `twisted.web.error` let's do that but I think combining that with this ticket would be doing too much at once.  (And as the comment says - this is already the way things were, mostly; just making it correct and consistent in the face of variable input, which we already had to deal with.)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

I think this is basically ready to land; I can appreciate Adi's concerns about comprehensibility, but this is a very small corner of the codebase, and mostly this is just describing the situation a bit better than it is now.  We can follow up with other fixes; nothing about those comments suggests to me that this change needs to be held up.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45857]) Merge downloadpage-py3-6197-2: Port twisted.web.client.downloadPage to Python 3

Author: hawkowl
Reviewers: adiroiban, glyph
Fixes: [#6197](https://github.com/twisted/twisted/issues/6197)
